tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py

```diff
@@ -54,24 +64,353 @@ def _load_hook_remove(
+            split_node.args = tuple(args)
+            ad_logger.debug(f"\nUpdated split node {split_node} arguments to {split_node.args}")
+
+
+def shard_weight_tensor(
+    gm: GraphModule,
+    weight_tensor: torch.Tensor,
+    param_key: str,
+    dim: int,
+    rank: int,
+    world_size: int,
+    min_local_shape: int = 1,
+    fused_weight_dims: Optional[list] = None,
+    custom_shard_fn: Optional[Callable[[torch.Tensor], torch.Tensor]] = None,
+    requires_grad: bool = False,
+    update_param: bool = True,
+) -> Tuple[torch.Tensor, torch.Size]:
+    """Shard a weight tensor across ranks and register load hook.
+
+    Args:
+        gm: GraphModule containing the weight
+        weight_tensor: The weight tensor to shard
+        param_key: Parameter key for registering load hook
+        dim: Dimension to shard along
+        rank: Current rank
+        world_size: Total number of ranks
+        min_local_shape: Minimum local shape constraint (for GQA)
+        fused_weight_dims: List of dimensions for fused weights
+        custom_shard_fn: Optional custom function to shard the tensor
+        requires_grad: Whether the parameter should require gradients
+        update_param: Whether to update the parameter in the module
+
+    Returns:
+        Tuple of (sharded_tensor, sharded_shape)
+    """
+
+    # Use custom shard function if provided
+    if custom_shard_fn is not None:
+        sharded_weight = custom_shard_fn(weight_tensor)
+        sharded_shape = sharded_weight.shape
+        # Register load hook with custom function
+        gm._register_load_state_dict_pre_hook(
+            partial(
+                _load_hook,
+                f_split=custom_shard_fn,
+                param_key=param_key,
+                param_shape=sharded_shape,
+            )
+        )
+
+    else:
+
+        def split_tensor(
+            t: torch.Tensor,
+            d: int = dim,
+            r: int = rank,
+            ws: int = world_size,
+            min_d_shape: int = min_local_shape,
+        ) -> torch.Tensor:
+            # The local tensor shape has to be divisible by min_d_shape
+            max_split_size = t.shape[d] // min_d_shape
+            if ws > max_split_size:
+                num_groups = math.ceil(ws / max_split_size)
+                ad_logger.debug(
+                    f"World size {ws} is greater than the max split size {max_split_size}. "
+                    + f"Splitting tensor to {num_groups} chunks"
+                )
+                return torch.tensor_split(t, max_split_size, dim=d)[r // num_groups]
+            return torch.tensor_split(t, ws, dim=d)[r]
+
+        # Handle fused weights
+        if fused_weight_dims is not None:
+            # Split fused weights, apply TP sharding to each, then concatenate back
+            sharded_weight = torch.cat(
+                [split_tensor(w) for w in torch.split(weight_tensor, fused_weight_dims, dim=dim)],
+                dim=dim,
+            )
```

Comment:
seems you should be able to just `f_split` as defined below to get the `sharded_weight` for either case